File: python/S4_BINARYSEARCH/L1_1D_ARRAY/1_binarysearch.py

class search:
    def __init__(self,arr):
        self.arr=arr
    def binarysearch(self,key):
        l=0
        h=len(self.arr)-1
        while(l<=h):
            mid=l+(h-l)//2
            if self.arr[mid]==key:
                return mid
            if self.arr[mid]>key:
                h=mid-1
            else :
                l=mid+1
        return -1
if __name__=="__main__":
    n=int(input("enter sizeof array: "))
    arr=list(map(int,input("enter element of array in sorted order :").split()))
    key=int(input("enter element to search :"))
    s=search(arr)
    r=s.binarysearch(key)
    if r==-1:
        print("element not found ")
    else:
        print("found at index :",r)

File: C/S4_BINARYSEARCH/L1_1D_ARRAY/1_binarysearch.c
#include<stdio.h>
int search(int arr[],int l,int r,int key)
{
    while(l<=r)
    {
        int mid=l+(r-l)/2;
        if(arr[mid]==key)return mid;
        if(arr[mid]>key)
        {
            r=mid-1;
        }
        else l=mid+1;
    }
    return -1;
}
int binarysearch2(int arr[],int l,int r,int key){
    if(l<r){
        int mid=l+(r-l)/2;
        if(arr[mid]==key){
            return 1;
        }
        if(arr[mid]>key){
            return binarysearch2(arr,l,mid,key);
        }
        else return binarysearch2(arr,mid+1,r,key);
    }
    return -1;
}
int main()
{
    int n;
    scanf("%d",&n);
    int arr[n];
    for(int i=0;i<n;i++)
    {
        scanf("%d",&arr[i]);
    }
    int key;
    scanf("%d",&key);
    int r=search(arr,0,n-1,key);
    if(r==-1)printf("not found");
    else printf("%d",r);
    return 0;

}